# Exercise #1 - All the photos!

### STEP 1.1

In [1]:
from bs4 import BeautifulSoup
import requests
from numpy import nan as NA
import numpy as np
import pandas as pd
pd.set_option('max_colwidth',150)

In [2]:
pokemon_url = 'https://pokemondb.net/pokedex/all'
# Issue a simple HTTP request to get the webpage text
pokemon_page = requests.get(pokemon_url)
# Response code is returned
pokemon_page

<Response [200]>

In [3]:
pokemon_page_soup = BeautifulSoup(pokemon_page.text, 'html.parser')

In [4]:
# Define a function that takes in a row of the pokedex table and returns it as a DataFrame with a single row. 
def pokemon_df(number):
    if number<=0:
        print('This pokemon does not exist.')
    else:
        # Extract the pokemon's name from the web
        pokemon=pokemon_page_soup.find_all(name="tr")[number]
        pokemon_values_list = pokemon.find_all(name="td")

        name_str = pokemon_values_list[1].text

        # Extract the pokemon's individual url address from the web
        url_str = pokemon.find(name="a", class_="ent-name").get('href')
        url_str = pokemon_url[0:21] + url_str


        # Extact the type(s) of the pokemon 
        type_str = pokemon_values_list[2].text

        # Extract the total points of the pokemon
        total_str = pokemon_values_list[3].text

        # The list of ID Number, HP, Attack, Defense, Sp. Atk, Sp. Def, Speed
        num_list=[]
        for i in range(7):
            num_list.append(pokemon.find_all(name="td", class_="cell-num")[i].text)
        # Return result as a dataframe created from a NP array
        pokemon_row = np.array([[ num_list[0],name_str, url_str, type_str, total_str,num_list[1],num_list[2],num_list[3],num_list[4],num_list[5],num_list[6]]])
        pokemon_df = pd.DataFrame(pokemon_row,columns = ['ID','NAME', 'URL', 'TYPE(s)', 'TOTAL', 'HP', 'Attack', 'Defense', 'Sp_Atk', 'Sp_Def', 'Speed'])
        return pokemon_df

In [5]:
pokemon_ids = [i+1 for i in range(1028)]
# Loop over our route IDs and append to the empty dataframe
# Store dataframes in a list
dfs = []
for id in pokemon_ids:
    pokemon_scrape_res = pokemon_df(id)
    # Append df to list
    dfs.append(pokemon_scrape_res)

In [6]:
pokemon_scrape_df = pd.concat(dfs)

In [7]:
# I have added the column names
# Convert str to numericals
pokemon_scrape_df.loc[:,'ID'] = pokemon_scrape_df.ID.astype(int)
pokemon_scrape_df.loc[:,'TOTAL'] = pokemon_scrape_df.TOTAL.astype(int)
pokemon_scrape_df.loc[:,'HP'] = pokemon_scrape_df.HP.astype(int)
pokemon_scrape_df.loc[:,'Attack'] = pokemon_scrape_df.Attack.astype(int)
pokemon_scrape_df.loc[:,'Defense'] = pokemon_scrape_df.Defense.astype(int)
pokemon_scrape_df.loc[:,'Sp_Atk'] = pokemon_scrape_df.Sp_Atk.astype(int)
pokemon_scrape_df.loc[:,'Sp_Def'] = pokemon_scrape_df.Sp_Def.astype(int)
pokemon_scrape_df.loc[:,'Speed'] = pokemon_scrape_df.Speed.astype(int)

In [8]:
# Remove duplicate values of pokemon based on the URL. 
result_df = pokemon_scrape_df.drop_duplicates(subset=['URL'], keep='first')

In [9]:
# Add a dummy variable to the DataFrame called "sample" that tags every 4th pokemon to be included in the sample. 
result_df=result_df.assign(sample1 = 0) 

In [10]:
for i in range(0,890):
    if i % 4 == 3:
        result_df.iat[i, -1] = 1

In [11]:
# Add a dummy variable to the DataFrame called "sample2" that tags every 12th pokemon to be included in the sample. 
result_df=result_df.assign(sample2 = 0) 

In [12]:
for i in range(0,890):
    if i % 12 == 11:
        result_df.iat[i, -1] = 1
result_df = result_df.set_index(['ID'])
result_df

NAME                                        URL  \
ID                                                                         
1                   Bulbasaur    https://pokemondb.net/pokedex/bulbasaur   
2                     Ivysaur      https://pokemondb.net/pokedex/ivysaur   
3                    Venusaur     https://pokemondb.net/pokedex/venusaur   
4                  Charmander   https://pokemondb.net/pokedex/charmander   
5                  Charmeleon   https://pokemondb.net/pokedex/charmeleon   
6                   Charizard    https://pokemondb.net/pokedex/charizard   
7                    Squirtle     https://pokemondb.net/pokedex/squirtle   
8                   Wartortle    https://pokemondb.net/pokedex/wartortle   
9                   Blastoise    https://pokemondb.net/pokedex/blastoise   
10                   Caterpie     https://pokemondb.net/pokedex/caterpie   
11                    Metapod      https://pokemondb.net/pokedex/metapod   
12                 Butterfree   https://pokemondb.net/pokedex/butterfree   
13                     Weedle       https://pokemondb.net/pokedex/weedle   
14                     Kakuna       https://pokemondb.net/pokedex/kakuna   
15                   Beedrill     https://pokemondb.net/pokedex/beedrill   
16                     Pidgey       https://pokemondb.net/pokedex/pidgey   
17                  Pidgeotto    https://pokemondb.net/pokedex/pidgeotto   
18                    Pidgeot      https://pokemondb.net/pokedex/pidgeot   
19                    Rattata      https://pokemondb.net/pokedex/rattata   
20                   Raticate     https://pokemondb.net/pokedex/raticate   
21                    Spearow      https://pokemondb.net/pokedex/spearow   
22                     Fearow       https://pokemondb.net/pokedex/fearow   
23                      Ekans        https://pokemondb.net/pokedex/ekans   
24                      Arbok        https://pokemondb.net/pokedex/arbok   
25                    Pikachu      https://pokemondb.net/pokedex/pikachu   
26                     Raichu       https://pokemondb.net/pokedex/raichu   
27                  Sandshrew    https://pokemondb.net/pokedex/sandshrew   
28                  Sandslash    https://pokemondb.net/pokedex/sandslash   
29                   Nidoran♀    https://pokemondb.net/pokedex/nidoran-f   
30                   Nidorina     https://pokemondb.net/pokedex/nidorina   
..                        ...                                        ...   
861                Grimmsnarl   https://pokemondb.net/pokedex/grimmsnarl   
862                 Obstagoon    https://pokemondb.net/pokedex/obstagoon   
863                Perrserker   https://pokemondb.net/pokedex/perrserker   
864                   Cursola      https://pokemondb.net/pokedex/cursola   
865                Sirfetch'd    https://pokemondb.net/pokedex/sirfetchd   
866                  Mr. Rime      https://pokemondb.net/pokedex/mr-rime   
867                 Runerigus    https://pokemondb.net/pokedex/runerigus   
868                   Milcery      https://pokemondb.net/pokedex/milcery   
869                  Alcremie     https://pokemondb.net/pokedex/alcremie   
870                   Falinks      https://pokemondb.net/pokedex/falinks   
871                Pincurchin   https://pokemondb.net/pokedex/pincurchin   
872                      Snom         https://pokemondb.net/pokedex/snom   
873                  Frosmoth     https://pokemondb.net/pokedex/frosmoth   
874               Stonjourner  https://pokemondb.net/pokedex/stonjourner   
875           Eiscue Ice Face       https://pokemondb.net/pokedex/eiscue   
876             Indeedee Male     https://pokemondb.net/pokedex/indeedee   
877   Morpeko Full Belly Mode      https://pokemondb.net/pokedex/morpeko   
878                    Cufant       https://pokemondb.net/pokedex/cufant   
879                Copperajah   https://pokemondb.net/pokedex/copperajah   
880                 Dracozolt    https://pokemondb.net/pokedex/dracozolt 

### STEP 1.2 & 1.3

In [13]:
from IPython.display import display
from IPython.display import Image

In [14]:
sample2=result_df[result_df['sample2'] == 1]

In [15]:
sample2.shape

(74, 12)

In [16]:
# Plots the pictures of the sample pokemons
for i in range(0,len(sample2)):
    # by changing index, the image of any pokemon can be shown
    pokemon_url = sample2.iat[i, 1]
    # Issue a simple HTTP request to get the webpage text
    pokemon_page = requests.get(pokemon_url)
    pokemon_page_soup = BeautifulSoup(pokemon_page.text, 'html.parser')
    # get imag urls for pokemons
    imag_url=pokemon_page_soup.find(name="img").get('src')
    display(Image(url=imag_url, unconfined=True))

# Exercise #2 - Rescrape and Clean the Location Table

### STEP 2.1

In [17]:
pokemon001_url =result_df.iat[0, 1]
pokemon001_url

'https://pokemondb.net/pokedex/bulbasaur'

In [18]:
# Extract the location table
tables = pd.read_html(requests.get(pokemon001_url, headers={'User-agent': 'Mozilla/5.0'}).text)

In [19]:
loc_df001=tables[-2]

In [20]:
# Some transformation of the table
loc_df001.columns = ["GAMES", "PLACE"]
loc_df001 = loc_df001.set_index(['GAMES'])
loc_df001=loc_df001.transpose()
loc_df001

GAMES      RedBlue         Yellow                GoldSilverCrystal  \
PLACE  Pallet Town  Cerulean City  Trade/migrate from another game   

GAMES                     RubySapphire FireRedLeafGreen  \
PLACE  Trade/migrate from another game      Pallet Town   

GAMES                          Emerald             DiamondPearlPlatinum  \
PLACE  Trade/migrate from another game  Trade/migrate from another game   

GAMES HeartGoldSoulSilver         BlackWhiteBlack 2White 2            XY  \
PLACE         Pallet Town  Trade/migrate from another game  Lumiose City   

GAMES         Omega RubyAlpha Sapphire                          SunMoon  \
PLACE  Trade/migrate from another game  Trade/migrate from another game   

GAMES Ultra SunUltra Moon  Let's Go PikachuLet's Go Eevee  \
PLACE             Route 2  Cerulean City, Viridian Forest   

GAMES                      SwordShield  
PLACE  Location data not yet available

In [21]:
# Get a list of locations where this pokemon at
gamesOrigin = loc_df001.columns.tolist()

In [22]:
# Get a list of game names for column name use
gamesNames = ['Red', 'Blue', 'Yellow', 'Gold', 'Silver', 'Crystal', 'Ruby', 'Sapphire', 'FireRed', 'LeafGreen', 'Emerald','Diamond','Pearl','Platinum','HeartGold','SoulSilver','Black','White','Black 2','White 2','X','Y','Omega Ruby','Alpha Sapphire','Sun','Moon','Ultra Sun','Ultra Moon',"Let's Go Pikachu","Let's Go Eevee",'Sword','Shield']

In [23]:
# Create two lists of index and content for future use
content=[]
for j in range(len(gamesNames)):
    i=0
    while i>=0:
        if gamesOrigin[i].find(gamesNames[j])!=-1:
            # When I search for Y game version, the index will goes to 1, since 'Yellow' has a 'Y' too, so I take it as a special case.
            if j==21 and gamesOrigin[i]=='Yellow':
                i+=1;
            else:
                content.append(loc_df001.iat[0,i])
                break;
        else:
            i=i+1
        if i==len(gamesOrigin):
            content.append(np.nan)
            break;

In [24]:
# Create the dateframe for 001
location001=pd.DataFrame(content).transpose()
location001.columns = gamesNames
location001.insert(0, "NAME", result_df.iat[0, 0], True)
location001.insert(0, "ID", result_df.index[0], True)
location001 = location001.set_index(['ID','NAME'])
location001

Red         Blue         Yellow  \
ID NAME                                                 
1  Bulbasaur  Pallet Town  Pallet Town  Cerulean City   

                                         Gold  \
ID NAME                                         
1  Bulbasaur  Trade/migrate from another game   

                                       Silver  \
ID NAME                                         
1  Bulbasaur  Trade/migrate from another game   

                                      Crystal  \
ID NAME                                         
1  Bulbasaur  Trade/migrate from another game   

                                         Ruby  \
ID NAME                                         
1  Bulbasaur  Trade/migrate from another game   

                                     Sapphire      FireRed    LeafGreen  ...  \
ID NAME                                                                  ...   
1  Bulbasaur  Trade/migrate from another game  Pallet Town  Pallet Town  ...   

                                   Omega Ruby  \
ID NAME                                         
1  Bulbasaur  Trade/migrate from another game   

                               Alpha Sapphire  \
ID NAME                                         
1  Bulbasaur  Trade/migrate from another game   

                                          Sun  \
ID NAME                                         
1  Bulbasaur  Trade/migrate from another game   

                                         Moon Ultra Sun Ultra Moon  \
ID NAME                                                              
1  Bulbasaur  Trade/migrate from another game   Route 2    Route 2   

                            Let's Go Pikachu                  Let's Go Eevee  \
ID NAME                                                                        
1  Bulbasaur  Cerulean City, Viridian Forest  Cerulean City, Viridian Forest   

                                        Sword                           Shield  
ID NAME                                                                         
1  Bulbasaur  Location data not yet available  Location data not yet available  

[1 rows x 32 columns]

### STEP 2.2

In [25]:
# Use sample 1 for this question
sample1=result_df[result_df['sample1'] == 1]

In [26]:
sample1.shape

(222, 12)

In [27]:
# Define a function that makes in a single pokemon's location table and returns it as a DataFrame with a single row. 
def pokemonLocation_df(ID):
    # Extract the pokemon's name from the web
    pokemon_url =result_df.iat[ID-1, 1]
    # Extract the location table
    tables = pd.read_html(requests.get(pokemon_url, headers={'User-agent': 'Mozilla/5.0'}).text)
    
    if ID >=807:
        loc_df = tables[-1]
    else:
        loc_df = tables[-2]
        
    loc_df.columns = ["GAMES", "PLACE"]
    loc_df = loc_df.set_index(['GAMES'])
    loc_df = loc_df.transpose()
    
    gamesOrigin=loc_df.columns.tolist()
    
    gamesNames = ['Red', 'Blue', 'Yellow', 'Gold', 'Silver', 'Crystal', 'Ruby', 'Sapphire', 'FireRed', 'LeafGreen', 'Emerald','Diamond','Pearl','Platinum','HeartGold','SoulSilver','Black','White','Black 2','White 2','X','Y','Omega Ruby','Alpha Sapphire','Sun','Moon','Ultra Sun','Ultra Moon',"Let's Go Pikachu","Let's Go Eevee",'Sword','Shield']
    content=[]
    for j in range(len(gamesNames)):
        i=0
        while i>=0:
            if gamesOrigin[i].find(gamesNames[j])!=-1:
                # When I search for Y game version, the index will goes to 1, since 'Yellow' has a 'Y' too, so I take it as a special case.
                if j==21 and gamesOrigin[i]=='Yellow':
                    i+=1
                else:
                    content.append(loc_df.iat[0,i])
                    break;
            else:
                i=i+1
            if i==len(gamesOrigin):
                content.append(np.nan)
                break;

    location = pd.DataFrame(content).transpose()
    location.columns = gamesNames
    location.insert(0, "NAME", result_df.iat[ID-1, 0], True)
    location.insert(0, "ID", result_df.index[ID-1], True)
    # Return result as a dataframe created from a NP array
    return location

In [28]:
# Loop over our sample IDs and append to the empty dataframe
dfs = []
for id in sample1.index:
    pokemon_scrape_res = pokemonLocation_df(id)
    #Append df to list
    dfs.append(pokemon_scrape_res)
pokemonLocations= pd.concat(dfs)
pokemonLocations = pokemonLocations.set_index(['ID'])
pokemonLocations.head()

NAME                                                    Red  \
ID                                                                      
4   Charmander                                            Pallet Town   
8    Wartortle                                        Evolve Squirtle   
12  Butterfree                                Evolve Caterpie/Metapod   
16      Pidgey  Route 1, 2, 3, 5, 6, 7, 8, 12, 13, 14, 15, 21, 24, 25   
20    Raticate                                   Route 16, 17, 18, 21   

                                                     Blue  \
ID                                                          
4                                             Pallet Town   
8                                         Evolve Squirtle   
12                                Evolve Caterpie/Metapod   
16  Route 1, 2, 3, 5, 6, 7, 8, 12, 13, 14, 15, 21, 24, 25   
20                                   Route 16, 17, 18, 21   

                                                             Yellow  \
ID                                                                    
4                                                          Route 24   
8                                                   Evolve Squirtle   
12                                          Evolve Caterpie/Metapod   
16  Route 1, 2, 5, 6, 7, 8, 11, 12, 13, 21, 24, 25, Viridian Forest   
20                     Route 9, 10, 11, 16, 18, 21, Pokémon Mansion   

                                                                                              Gold  \
ID                                                                                                   
4                                                                  Trade/migrate from another game   
8                                                                  Trade/migrate from another game   
12  Route 2, 26, 27, 34, 35, 36, 37, 38, 39, Azalea Town, Ilex Forest, Lake of Rage, National Park   
16                                     Route 1, 2, 5, 6, 25, 29, 30, 31, 35, 36, 37, National Park   
20               Route 7, 9, 10, 26, 27, 38, 39, Burned Tower, Mt. Mortar, Tohjo Falls, Union Cave   

                                                                               Silver  \
ID                                                                                      
4                                                     Trade/migrate from another game   
8                                                     Trade/migrate from another game   
12                                                                      National Park   
16                        Route 1, 2, 5, 6, 25, 29, 30, 31, 35, 36, 37, National Park   
20  Route 7, 9, 10, 26, 27, 38, 39, Burned Tower, Mt. Mortar, Tohjo Falls, Union Cave   

                                                                                                          Crystal  \
ID                                                                                                                  
4                                                                                 Trade/migrate from another game   
8                                                                                 Trade/migrate from another game   
12                                                                    Route 2, 24, 25, Ilex Forest, National Park   
16                                  Route 1, 2, 5, 25, 29, 30, 31, 32, 34, 35, 36, 37, Ilex Forest, National Park   
20  Route 1, 3, 4, 6, 7, 9, 10, 11, 21, 26, 27, 38, 39, 42, 43, Burned Tower, Mt. Mortar, Tohjo Falls, Union Cave   

                               Ruby                         Sapphire  \
ID                                                                     
4   Trade/migrate from another game  Trade/migrate from another game   
8   Trade/migrate from another game  Trade/migrate from another game   
12  Trade/migrate from another game  Trade/migrate from another game   
16  Trade/migrate from another game  Trade/migrate fr

# Exercise #3 - Location Analysis Done Right

### STEP 3.1

In [29]:
# Join the pokedex data to the  location DataFrame
pokemonStat=pokemonLocations.reset_index().merge(result_df, how="left").set_index(pokemonLocations.index.names)
pokemonStat

NAME  \
ID                          
4              Charmander   
8               Wartortle   
12             Butterfree   
16                 Pidgey   
20               Raticate   
24                  Arbok   
28              Sandslash   
32               Nidoran♂   
36               Clefable   
40             Wigglytuff   
44                  Gloom   
48                Venonat   
52                 Meowth   
56                 Mankey   
60                Poliwag   
64                Kadabra   
68                Machamp   
72              Tentacool   
76                  Golem   
80                Slowbro   
84                  Doduo   
88                 Grimer   
92                 Gastly   
96                Drowzee   
100               Voltorb   
104                Cubone   
108             Lickitung   
112                Rhydon   
116                Horsea   
120                Staryu   
..                    ...   
772            Type: Null   
776            Turtonator   
780                Drampa   
784               Kommo-o   
788             Tapu Fini   
792                Lunala   
796             Xurkitree   
800              Necrozma   
804             Naganadel   
808                Meltan   
812             Rillaboom   
816                Sobble   
820              Greedent   
824               Blipbug   
828               Thievul   
832               Dubwool   
836               Boltund   
840                Applin   
844            Sandaconda   
848                 Toxel   
852             Clobbopus   
856               Hatenna   
860               Morgrem   
864               Cursola   
868               Milcery   
872                  Snom   
876         Indeedee Male   
880             Dracozolt   
884             Duraludon   
888  Zacian Crowned Sword   

                                                                                                                                                       Red  \
ID                                                                                                                                                           
4                                                                                                                                              Pallet Town   
8                                                                                                                                          Evolve Squirtle   
12                                                                                                                                 Evolve Caterpie/Metapod   
16                                                                                                   Route 1, 2, 3, 5, 6, 7, 8, 12, 13, 14, 15, 21, 24, 25   
20                                                                                                                                    Route 16, 17, 18, 21   
24                                                                                                                                 Route 23, Cerulean Cave   
28                                                                                                                         Trade/migrate from another game   
32                                                                                                                                   Route 22, Safari Zone   
36                                                                                                                                         Evolve Clefairy   
40                                                                                                                                           Cerulean Cave   
44                                                                                                                                    Route 12, 13, 14, 15   
48                                                                                                                       Route 12, 13, 14, 

In [30]:
pokemonX=pokemonStat[['NAME','X','TOTAL']]
pokemonX=pokemonX[pokemonX.index<= 720]
pokemonX

NAME  \
ID                          
4              Charmander   
8               Wartortle   
12             Butterfree   
16                 Pidgey   
20               Raticate   
24                  Arbok   
28              Sandslash   
32               Nidoran♂   
36               Clefable   
40             Wigglytuff   
44                  Gloom   
48                Venonat   
52                 Meowth   
56                 Mankey   
60                Poliwag   
64                Kadabra   
68                Machamp   
72              Tentacool   
76                  Golem   
80                Slowbro   
84                  Doduo   
88                 Grimer   
92                 Gastly   
96                Drowzee   
100               Voltorb   
104                Cubone   
108             Lickitung   
112                Rhydon   
116                Horsea   
120                Staryu   
..                    ...   
604            Eelektross   
608               Lampent   
612               Haxorus   
616               Shelmet   
620              Mienshao   
624              Pawniard   
628              Braviary   
632                Durant   
636              Larvesta   
640              Virizion   
644                Zekrom   
648   Meloetta Aria Forme   
652            Chesnaught   
656               Froakie   
660             Diggersby   
664            Scatterbug   
668                Pyroar   
672                Skiddo   
676               Furfrou   
680              Doublade   
684               Swirlix   
688               Binacle   
692             Clauncher   
696                Tyrunt   
700               Sylveon   
704                 Goomy   
708              Phantump   
712              Bergmite   
716               Xerneas   
720  Hoopa Hoopa Confined   

                                                                                                      X  \
ID                                                                                                        
4                                                                                          Lumiose City   
8                                                                                       Evolve Squirtle   
12                                                                              Evolve Caterpie/Metapod   
16                                                                                           Route 2, 3   
20                                                                      Trade/migrate from another game   
24                                                                                             Route 19   
28                                                                              Route 18, Terminus Cave   
32                                                                                             Route 11   
36                                                                      Trade/migrate from another game   
40                                                                          Evolve Jigglypuff/Igglybuff   
44                                                                                        Evolve Oddish   
48                                                                      Trade/migrate from another game   
52                                                                      Trade/migrate from another game   
56                                                                      Trade/migrate from another game   
60   Route 14, 15, 16, 19, 21, Laverre City, Frost Cavern, Couriway Town, Pokémon Village, Victory Road   
64                                                                                          Evolve Abra   
68                                                                                Evolve Machop/Machoke   
72                                    Route 8, 12, Cyllage City, Ambrette Town, Shalour City, Azure Bay   
76                                                   

In [31]:
def explode_str(df, col, sep):
    s = df[col]
    i = np.arange(len(s)).repeat(s.str.count(sep) + 1)
    return df.iloc[i].assign(**{col: sep.join(s).split(sep)})

In [32]:
# separate different locations in one cell
pokemonX=explode_str(pokemonX, 'X', ',')
pokemonX

NAME                                X  TOTAL
ID                                                               
4              Charmander                     Lumiose City    309
8               Wartortle                  Evolve Squirtle    405
12             Butterfree          Evolve Caterpie/Metapod    395
16                 Pidgey                          Route 2    251
16                 Pidgey                                3    251
20               Raticate  Trade/migrate from another game    413
24                  Arbok                         Route 19    448
28              Sandslash                         Route 18    450
28              Sandslash                    Terminus Cave    450
32               Nidoran♂                         Route 11    273
36               Clefable  Trade/migrate from another game    483
40             Wigglytuff      Evolve Jigglypuff/Igglybuff    435
44                  Gloom                    Evolve Oddish    395
48                Venonat  Trade/migrate from another game    305
52                 Meowth  Trade/migrate from another game    290
56                 Mankey  Trade/migrate from another game    305
60                Poliwag                         Route 14    300
60                Poliwag                               15    300
60                Poliwag                               16    300
60                Poliwag                               19    300
60                Poliwag                               21    300
60                Poliwag                     Laverre City    300
60                Poliwag                     Frost Cavern    300
60                Poliwag                    Couriway Town    300
60                Poliwag                  Pokémon Village    300
60                Poliwag                     Victory Road    300
64                Kadabra                      Evolve Abra    400
68                Machamp            Evolve Machop/Machoke    505
72              Tentacool                          Route 8    335
72              Tentacool                               12    335
..                    ...                              ...    ...
632                Durant                    Terminus Cave    484
636              Larvesta  Trade/migrate from another game    360
640              Virizion  Trade/migrate from another game    580
644                Zekrom  Trade/migrate from another game    680
648   Meloetta Aria Forme  Trade/migrate from another game    600
652            Chesnaught         Evolve Chespin/Quilladin    530
656               Froakie                   Vaniville Town    314
656               Froakie                   Aquacorde Town    314
660             Diggersby                         Route 22    423
664            Scatterbug                          Route 2    200
664            Scatterbug                 Santalune Forest    200
668                Pyroar                    Evolve Litleo    507
672                Skiddo                          Route 5    350
676               Furfrou                          Route 5    472
680              Doublade                   Evolve Honedge    448
684               Swirlix                          Route 7    341
688               Binacle                          Route 8    306
688               Binacle                               12    306
688               Binacle                    Ambrette Town    306
688               Binacle                        Azure Bay    306
692             Clauncher                          Route 8    330
692             Clauncher                     Cyllage City    330
692             Clauncher                    Ambrette Town    330
696                Tyrunt                    Ambrette Town    362
700               Sylveon                     Evolve Eevee    525
704                 Goomy                         Route 14    300
708              Phantump                         Route 16    309
712              Bergmite                     Frost Cavern   

In [33]:
#For the locations in pokemon X, calculate the average total points for each location. 
pokemonX.groupby(by='X')['TOTAL'].mean().sort_values(ascending=False)

X
Team Flare HQ                 680.0
 Roaming Kalos                580.0
Sea Spirit's Den              580.0
Evolve Axew/Fraxure           540.0
Evolve Vanillite/Vanillish    535.0
Evolve Rhyhorn/Rhydon         535.0
Evolve Chespin/Quilladin      530.0
Evolve Eevee                  525.0
Evolve Nosepass               525.0
Tower of Mastery              525.0
Evolve Gligar                 510.0
Evolve Mienfoo                510.0
Evolve Litleo                 507.0
Evolve Machop/Machoke         505.0
Evolve Scyther                500.0
Evolve Pansage                498.0
Evolve Panpour                498.0
Evolve Geodude/Graveler       495.0
Evolve Anorith                495.0
Route 17                      494.0
Evolve Slowpoke               490.0
Evolve Gothita/Gothorita      490.0
Evolve Scraggy                488.0
Evolve Rhyhorn                485.0
Route 12                      485.0
Evolve Clamperl               485.0
Evolve Lotad/Lombre           480.0
Evolve Combee             

Team Flare HQ has the highest average total point score at 680.

# END